src/rdb.c

```diff
@@ -3030,6 +3030,19 @@ int rdbLoadRioWithLoadingCtx(rio *rdb, int rdbflags, rdbSaveInfo *rsi, rdbLoadin
                 exit(1);
             }
             db = rdb_loading_ctx->dbarray+dbid;
+            /* Cluster enabled nodes MUST maintain empty non zero databases as
+             * valid state. Otherwise, a key stored in non zero database becomes
+             * a poison pill that can replicate to other nodes, be SAVEed to RDB
+             * files, AND crash nodes trying to restore from that RDB.
+             * Crashing early gives a chance to detect, investigate and possibly
+             * save the offending data without propagating the error.
+             */
+            if (dbid != 0 && server.cluster_enabled) {
+                serverLog(LL_WARNING,
+                    "You can't have keys in a DB different than DB 0 when in "
+                    "Cluster mode. Exiting.");
+                exit(1); 
```

Comment:
Which is better to ignore db other than db 0 or start failure?

Reply:
I think better to assert/exit when we face this poison pill rather than silently lose data

Reply:
how will that situation ever occur?
if it's a real possibility due to some innocent mistake, then maybe we should warn and not exit.
if it would never really occur, then i'm not sure we need it at all (let alone that big comment).
if it can occur due to some abuse, then a warning and exit seems right.

however, why isn't the previous check in that block (`dbid >= (unsigned)server.dbnum`) enough?
maybe the problem is that for some reason redis starts in cluster mode and has more than one db configured?

Reply:
I agree that this can probably we solved by existing configuration for databases.
I think, if I recall correctly, that in this case a cluster mode enabled node was provided a cluster mode disabled RDB snapshot. By maybe @llahav-amzn has a better example case.

Reply:
i see it is possible to start a redis server with `cluster-enabled` set to `yes` and `databases` set to more than one db.
SELECT would be rejected, but apparently that's not enough.
i guess that's the problem here... we don't need another check in rdb.c, instead we need to either reject this configuration with a startup error, or silently set `databases` to `1` on startup.

Reply:
@oranagra So what do you think is the best way to progress?
AFAIK the main motivation here was only to place a more defensive code, rather than to face a harder to debug problem.
We can alternatively make the CME configuration to force databases value to be set to 1.

Reply:
i think we have two non-mutable configs that are mutually exclusive.
i think redis should refuse to start with cluster enabled and more than one db.
i'm not sure why it wasn't like that from the beginning, maybe i'm missing something.
doing that, will implicitly solve the rdb loading problem.